### DISCUSSION TITLE
only 4.3.0 working

### DISCUSSION THREAD

**Discussion by Turro75** [vintagepc/MINI404#55] [General]
First of all Thanks a lot for Your work, impressive!

I got it working in Ubuntu 20.04.2 LTS 64bit.
I have been able to run only 4.3.0 firmware taken from prusa buddy github.
Not able to run neither a stock 4.2.1 (stucks at the end of verify then it resets) nor the attached 4.3.2 compiled and customized by me (bsod)

![Schermata da 2021-03-14 09-26-57](https://user-images.githubusercontent.com/10894566/111062146-7e66c000-84a7-11eb-8aa7-cce1a3b60c90.png)

![Schermata da 2021-03-14 09-29-38](https://user-images.githubusercontent.com/10894566/111062212-d43b6800-84a7-11eb-8602-cdd771678219.png)

[firmware.bbf.zip](https://github.com/vintagepc/MINI404/files/6136286/firmware.bbf.zip)

Valerio

**Comment by vintagepc** [comment]
Hi there! Glad it is (mostly) working for you!

There's a known quirk in the 4.2.1 firmware that makes it slow to start (It expects the TMC drivers to reply to write requests over serial when they don't actually, at least according to the datasheet). But it is entirely possible that a recent change has broken something in the simulation. I will check.

In regards to your custom 4.3.2, I do not know what changes you made so I can't really say what might be wrong. Without your sources the best advice I can offer here is to use gdb debugging of the firmware (See wiki/Getting-Started for info how to debug with vscode or CLI gdb) and breakpoint the bsod() function in appmain.cpp - then you can inspect the stack and see what is causing the BSOD.  I would also suggest using the _debug_noboot .bin file when testing non-official bbf files just to eliminate any undiscovered bugs relating to signature verification.

**Reply by Turro75** [reply]
I modified Your PR to get uart6 working, if You remember on a real buddy the read from uart was not working with your PR. so I changed the DMA IT code.
here my diff file from prusa sources [https://drive.google.com/file/d/1Z0si_asZ3RmFuyAJANE-2SG_g1kRgSkn/view?usp=sharing](https://drive.google.com/file/d/1Z0si_asZ3RmFuyAJANE-2SG_g1kRgSkn/view?usp=sharing)

**Reply by vintagepc** [reply]
Ah, I see. DMA is still somewhat fragile (pending #30) and requires some internal handling to specify data addresses for peripherals; that is probably the cause of the issue with your custom firmware as I do not think DMA support was added to UART6. I can't make any guarantees as to when, but I will take a look and see if it is an easy fix.

**Reply by vintagepc** [reply]
The BBF you provided does not seem to be the same as the diff. I just built RELEASE_4.3.1 firmware.bin with your diff applied and it boots. Serial output appears on the correct USART; no BSOD. 

Input does not work and normal execution throws "command overflow" errors, which are odd - those come from the uartslave_cycle() call, which uses main.cpp::uart6slave. However, your #ifdefs for USE_UART_6_SERIAL around line 240 mean this item is not initialized anymore and you are accessing uninitialized memory. 

I don't see anything that suggests these changes expose a problem in the simulator at the moment and unfortunately I cannot tell whether there actually is one until the "uninitialized memory" problem is resolved.

**Reply by Turro75** [reply]
thanks for the tips, anyway even commenting that function call I'm still getting bsod in the emulator.
the firmware.bbf above is compiled on 4.3.2 branch

**Reply by vintagepc** [reply]
Ah, ok - that explains it. I'd just assumed you'd bumped the version yourself instead of an official 4.3.2. The 4.3.2 RC has some HAL/USB Changes to the USB CDC setup, which is still unimplemented (Qemu does not have a facility I am aware of for allowing a host to use a guest as a usb device) For now this can be worked around by disabling the MX_USB_DEVICE_Init(); call in main.cpp (~line 976).

I will look at implementing the hardware sufficiently that it boots again with 4.3.2 without any firmware changes. Sorry for the confusion.

**Reply by vintagepc** [reply]
Looks like it is satisfactory to stub out the FS device registers with the HS implementation temporarily. I'm now able to start your provided .bbf file without a BSOD.

![image](https://user-images.githubusercontent.com/53943260/111081585-63617380-84da-11eb-9ef2-a44ff55addb1.png)

It will be resolved with commit b14be125b8d86fbd10864b64ace9e721bb1cc8a7 when #56  is merged.

**Reply by Turro75** [reply]
I confirm it is working now. Thanks

**Reply by vintagepc** [reply]
Fantastic, glad I could help!

**Comment by vintagepc** [comment]
OK, I think I see what is going on with 4.2.1 - if using the official .bbf the bootloader sets up the watchdog and causes a watchdog reset due to the long wait time before the TMC serial read fails. I suspect if one uses a self-compiled 4.2.1 mini_debug_noboot.bin file instead then it will work around this issue.
